src/unix/thread.c

```diff
@@ -60,51 +116,62 @@ sysreturn clone(unsigned long flags, void *child_stack, int *ptid, unsigned long
     thread_log(current, "clone: flags %lx, child_stack %p, ptid %p, ctid %p, newtls %lx",
         flags, child_stack, ptid, ctid, newtls);
 
-    if (!(flags & CLONE_THREAD)) {
-        thread_log(current, "attempted to create new process, aborting.");
-        return set_syscall_error(current, ENOSYS);
-    }
-
-    /* no stack size given, just validate the top word */
-    if (!validate_user_memory(child_stack, sizeof(u64), true))
-        return set_syscall_error(current, EFAULT);
-
-    if (((flags & CLONE_PARENT_SETTID) &&
-         !validate_user_memory(ptid, sizeof(int), true)) ||
-        ((flags & CLONE_CHILD_CLEARTID) &&
-         !validate_user_memory(ctid, sizeof(int), true)))
-        return set_syscall_error(current, EFAULT);
-
-    thread t = create_thread(current->p, INVALID_PHYSICAL);
-    context_frame f = thread_frame(t);
-    /* clone frame processor state */
-    clone_frame_pstate(f, thread_frame(current));
-    thread_clone_sigmask(t, current);
-
-    /* clone behaves like fork at the syscall level, returning 0 to the child */
-    set_syscall_return(t, 0);
-    f[SYSCALL_FRAME_SP] = u64_from_pointer(child_stack);
-    if (flags & CLONE_SETTLS)
-        set_tls(f, newtls);
-    if (flags & CLONE_PARENT_SETTID)
-        *ptid = t->tid;
-    if (flags & CLONE_CHILD_CLEARTID)
-        t->clear_tid = ctid;
-    t->blocked_on = 0;
-    t->syscall = 0;
-    f[FRAME_FULL] = true;
-    thread_reserve(t);
-    schedule_thread(t);
-    return t->tid;
+    struct clone_args_internal args = {
+         .flags = flags,
+         .child_tid = ctid,
+         .parent_tid = ptid,
+         /* no stack size given, just validate the top word */
+         .stack = child_stack - sizeof(u64),
+         .stack_size = sizeof(u64),
+         .tls = newtls,
+    };
+
+    return clone_internal(&args);
 }
 
+struct clone_args {
```

Comment:
Structs defined by the syscall interface typically go in ```src/unix/system_structs.h```, if arch-independent, or ```src/<arch>/unix_machine.h``` otherwise.